File: animator/ui/src/ai/snapshot.ts
// ===========================================================================
// AI SNAPSHOT — read-only, tiered, aliased, revision-stamped scene view
// (A3 / E-AI-2 / spec 06 + AI-REQ-001/003).
//
// The engine (snapshot.rs) returns compact truth. THIS layer adds what the
// prompt pipeline needs: defensive parsing, deterministic aliases (n1/l1/s1 —
// the model never memorizes raw ids and can never inject them, spec 06/15),
// by-ref lookup, and a token-bounded textual render. The view is frozen —
// code cannot mutate a document through a snapshot (AI-REQ-003).
//
// Staleness: `rev` comes from the engine's E-AI-4 counter. The orchestrator
// (A6) rebuilds a view per turn and after every transaction; the validator's
// apply-time checks (A4/A5) always run against LIVE state — this view is
// advisory, never authority.
// ===========================================================================

// ---------------------------------------------------------------------------
// Rows (mirrors snapshot.rs; unknown extra fields are ignored, missing fields
// get safe defaults — fail-closed on format/version mismatch instead).
// ---------------------------------------------------------------------------

export interface SnapSettingsRow {
  w: number
  h: number
  fps: number
  bg: string
  bgA: number
}

export interface SnapKfRow {
  f: number
  label?: string
  blank?: boolean
  n?: number
}

export interface SnapTweenRow {
  s: number
  e: number
  ease: number
}

export interface SnapLayerRow {
  i: number
  id: number
  name: string
  kind: string
  vis: boolean
  lock: boolean
  outline?: boolean
  oc?: string
  parent?: number
  collapsed?: boolean
  kf: SnapKfRow[]
  tw: SnapTweenRow[]
}

export interface SnapNodeRow {
  id: number
  kind: string
  /** [layerIndex, frame] membership pairs. */
  kf: Array<[number, number]>
  x?: number
  y?: number
  sx?: number
  sy?: number
  r?: number
  w?: number
  h?: number
  fill?: string
  stroke?: string
  sw?: number
  sym?: number
  lp?: string
  ff?: number
}

export interface SnapSymbolRow {
  id: number
  name: string
  type: string
  uses: number
  dur: number
}

export interface SceneSnapshot {
  v: number
  rev: number
  settings: SnapSettingsRow
  scene: { i: number; name: string; count: number }
  activeLayer: number
  playhead: number
  duration: number
  selection: number[]
  counts: { layers: number; nodes: number; keyframes: number; tweens: number; symbols: number }
  layers: SnapLayerRow[]
  nodes: SnapNodeRow[]
  library: SnapSymbolRow[]
}

export class SnapshotError extends Error {
  readonly code = 'E_SNAPSHOT'
  constructor(message: string) {
    super(message)
    this.name = 'SnapshotError'
  }
}

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v)
}

function numOr(v: unknown, d: number): number {
  return typeof v === 'number' && Number.isFinite(v) ? v : d
}

function strOr(v: unknown, d: string): string {
  return typeof v === 'string' ? v : d
}

function boolOr(v: unknown, d: boolean): boolean {
  return typeof v === 'boolean' ? v : d
}

/** Parse + normalize engine snapshot JSON. Throws SnapshotError on garbage. */
export function parseSceneSnapshot(json: string): SceneSnapshot {
  let raw: unknown
  try {
    raw = JSON.parse(json)
  } catch {
    throw new SnapshotError('snapshot JSON parse fail')
  }
  if (!isRecord(raw)) throw new SnapshotError('snapshot is not an object')
  if (raw.v !== 1) throw new SnapshotError(`unsupported snapshot version: ${String(raw.v)}`)

  const settings = isRecord(raw.settings) ? raw.settings : {}
  const scene = isRecord(raw.scene) ? raw.scene : {}
  const counts = isRecord(raw.counts) ? raw.counts : {}

  const layers: SnapLayerRow[] = Array.isArray(raw.layers)
    ? raw.layers.filter(isRecord).map((l) => ({
        i: numOr(l.i, 0),
        id: numOr(l.id, 0),
        name: strOr(l.name, ''),
        kind: strOr(l.kind, 'normal'),
        vis: boolOr(l.vis, true),
        lock: boolOr(l.lock, false),
        ...(l.outline === true ? { outline: true, oc: strOr(l.oc, '#ff0000') } : {}),
        ...(typeof l.parent === 'number' ? { parent: l.parent } : {}),
        ...(l.collapsed === true ? { collapsed: true } : {}),
        kf: Array.isArray(l.kf)
          ? l.kf.filter(isRecord).map((k) => ({
              f: numOr(k.f, 0),
              ...(typeof k.label === 'string' ? { label: k.label } : {}),
              ...(k.blank === true ? { blank: true } : {}),
              ...(typeof k.n === 'number' ? { n: numOr(k.n, 0) } : {}),
            }))
          : [],
        tw: Array.isArray(l.tw)
          ? l.tw
              .filter(isRecord)
              .map((t) => ({ s: numOr(t.s, 0), e: numOr(t.e, 0), ease: numOr(t.ease, 0) }))
          : [],
      }))
    : []

  const nodes: SnapNodeRow[] = Array.isArray(raw.nodes)
    ? raw.nodes.filter(isRecord).map((n) => {
        const row: SnapNodeRow = {
          id: numOr(n.id, 0),
          kind: strOr(n.kind, 'rect'),
          kf: Array.isArray(n.kf)
            ? n.kf
                .filter((p): p is [unknown, unknown] => Array.isArray(p) && p.length === 2)
                .map((p) => [numOr(p[0], 0), numOr(p[1], 0)] as [number, number])
            : [],
        }
        for (const key of ['x', 'y', 'sx', 'sy', 'r', 'w', 'h', 'sw', 'sym', 'ff'] as const) {
          if (typeof n[key] === 'number') row[key] = n[key] as number
        }
        if (typeof n.fill === 'string') row.fill = n.fill
        if (typeof n.stroke === 'string') row.stroke = n.stroke
        if (typeof n.lp === 'string') row.lp = n.lp
        return row
      })
    : []

  const library: SnapSymbolRow[] = Array.isArray(raw.library)
    ? raw.library.filter(isRecord).map((s) => ({
        id: numOr(s.id, 0),
        name: strOr(s.name, ''),
        type: strOr(s.type, 'graphic'),
        uses: numOr(s.uses, 0),
        dur: numOr(s.dur, 1),
      }))
    : []

  return {
    v: 1,
    rev: numOr(raw.rev, 0),
    settings: {
      w: numOr(settings.w, 1920),
      h: numOr(settings.h, 1080),
      fps: numOr(settings.fps, 24),
      bg: strOr(settings.bg, '#ffffff'),
      bgA: numOr(settings.bgA, 1),
    },
    scene: {
      i: numOr(scene.i, 0),
      name: strOr(scene.name, 'Scene 1'),
      count: numOr(scene.count, 1),
    },
    activeLayer: numOr(raw.active_layer, 0),
    playhead: numOr(raw.playhead, 1),
    duration: numOr(raw.duration, 1),
    selection: Array.isArray(raw.selection)
      ? raw.selection.filter((x): x is number => typeof x === 'number')
      : [],
    counts: {
      layers: numOr(counts.layers, 0),
      nodes: numOr(counts.nodes, 0),
      keyframes: numOr(counts.keyframes, 0),
      tweens: numOr(counts.tweens, 0),
      symbols: numOr(counts.symbols, 0),
    },
    layers,
    nodes,
    library,
  }
}

// ---------------------------------------------------------------------------
// View: aliases + lookup + frozen read-only access
// ---------------------------------------------------------------------------

export interface SceneSnapshotView {
  /** E-AI-4 revision at build time (staleness checks). */
  readonly rev: number
  /** Frozen parsed snapshot. */
  readonly raw: SceneSnapshot
  /** Deterministic alias for a node ('n1'…) / layer ('l1'…) / symbol ('s1'…). */
  aliasOf(id: number, kind: 'n' | 'l' | 's'): string | undefined
  /** Reverse: alias back to the real id (never trust model-produced ids). */
  idOf(alias: string): number | undefined
  /** Lookup by real id OR alias. */
  node(ref: number | string): SnapNodeRow | undefined
  layer(ref: number | string): SnapLayerRow | undefined
  symbol(ref: number | string): SnapSymbolRow | undefined
  /** Token-bounded text render for the prompt (A6 PromptBuilder consumes). */
  toPromptText(): string
}

function trunc(name: string): string {
  return name.length > 40 ? `${name.slice(0, 37)}…` : name
}

export function buildSnapshotView(json: string): SceneSnapshotView {
  const raw = parseSceneSnapshot(json)

  const aliasToId = new Map<string, number>()
  const idToAlias = new Map<string, string>() // `${kind}${id}` → alias
  raw.nodes.forEach((n, i) => {
    const alias = `n${i + 1}`
    aliasToId.set(alias, n.id)
    idToAlias.set(`n${n.id}`, alias)
  })
  raw.layers.forEach((l, i) => {
    const alias = `l${i + 1}`
    aliasToId.set(alias, l.id)
    idToAlias.set(`l${l.id}`, alias)
  })
  raw.library.forEach((s, i) => {
    const alias = `s${i + 1}`
    aliasToId.set(alias, s.id)
    idToAlias.set(`s${s.id}`, alias)
  })

  // Freeze the whole graph: a snapshot has NO write path (AI-REQ-003).
  const freezeDeep = <T>(v: T): T => {
    if (Array.isArray(v)) {
      v.forEach(freezeDeep)
      Object.freeze(v)
    } else if (typeof v === 'object' && v !== null) {
      Object.values(v).forEach(freezeDeep)
      Object.freeze(v)
    }
    return v
  }
  freezeDeep(raw)

  function resolve(ref: number | string, fallbackKind: 'n' | 'l' | 's'): number | undefined {
    if (typeof ref === 'number') return ref
    const id = aliasToId.get(ref)
    if (id !== undefined) return id
    return aliasToId.get(`${fallbackKind}${ref}`)
  }

  return {
    rev: raw.rev,
    raw,
    aliasOf(id, kind) {
      return idToAlias.get(`${kind}${id}`)
    },
    idOf(alias) {
      return aliasToId.get(alias)
    },
    node(ref) {
      const id = resolve(ref, 'n')
      return raw.nodes.find((n) => n.id === id)
    },
    layer(ref) {
      if (typeof ref === 'number') {
        // Bare numbers are layer INDICES (matching snapshot rows + session fns).
        return raw.layers.find((l) => l.i === ref)
      }
      const id = resolve(ref, 'l')
      return raw.layers.find((l) => l.id === id)
    },
    symbol(ref) {
      const id = resolve(ref, 's')
      return raw.library.find((s) => s.id === id)
    },
    toPromptText(): string {
      const s = raw.settings
      const lines: string[] = []
      lines.push(
        `KINEORA SNAPSHOT rev=${raw.rev} | stage ${s.w}x${s.h} @${s.fps}fps bg=${s.bg}` +
          ` | scene[${raw.scene.i}] "${trunc(raw.scene.name)}" (${raw.scene.count} scenes)` +
          ` | playhead=${raw.playhead} duration=${raw.duration}`,
      )
      const selAliases = raw.selection
        .map((id) => idToAlias.get(`n${id}`) ?? `#${id}`)
        .join(' ')
      lines.push(`selection: ${selAliases || '(none)'}`)
      for (const l of raw.layers) {
        const la = idToAlias.get(`l${l.id}`) ?? `l${l.i + 1}`
        const flags = [
          l.kind !== 'normal' ? l.kind : null,
          l.vis ? null : 'hidden',
          l.lock ? 'locked' : null,
          l.outline ? 'outline' : null,
          l.collapsed ? 'collapsed' : null,
        ]
          .filter(Boolean)
          .join(',')
        const kf = l.kf
          .map((k) => `${k.f}${k.blank ? '(blank)' : ''}${k.label ? ` "${trunc(k.label)}"` : ''}`)
          .join(',')
        const tw = l.tw.map((t) => `${t.s}..${t.e} ease=${t.ease}`).join(';')
        lines.push(
          `layer ${la} "${trunc(l.name)}"${flags ? ` [${flags}]` : ''} kf:${kf || '-'} tw:${tw || '-'}`,
        )
      }
      for (const n of raw.nodes) {
        const na = idToAlias.get(`n${n.id}`) ?? `n${n.id}`
        const where = n.kf.map(([li, f]) => `l${li + 1}@${f}`).join(',')
        if (n.kind === 'symbol') {
          const sa = n.sym !== undefined ? (idToAlias.get(`s${n.sym}`) ?? `s${n.sym}`) : '?'
          lines.push(
            `node ${na} symbol→${sa} at (${n.x},${n.y}) s(${n.sx},${n.sy}) r${n.r} on ${where}`,
          )
        } else {
          lines.push(
            `node ${na} ${n.kind} fill=${n.fill ?? '-'} stroke=${n.stroke ?? 'none'}/${n.sw ?? '-'}` +
              ` size=${n.w}x${n.h} at (${n.x},${n.y}) s(${n.sx},${n.sy}) r${n.r} on ${where}`,
          )
        }
      }
      if (raw.library.length > 0) {
        lines.push(
          `library: ${raw.library
            .map((sy) => `${idToAlias.get(`s${sy.id}`) ?? sy.id} "${trunc(sy.name)}" ${sy.type} uses=${sy.uses} dur=${sy.dur}`)
            .join(' | ')}`,
        )
      }
      lines.push(
        `counts: layers=${raw.counts.layers} nodes=${raw.counts.nodes} keyframes=${raw.counts.keyframes} tweens=${raw.counts.tweens} symbols=${raw.counts.symbols}`,
      )
      return lines.join('\n')
    },
  }
}

File: animator/core/src/snapshot.rs
//! E-AI-2 + E-AI-5 (A3, D-0010) — compact semantic scene snapshot and the
//! trusted runtime capability manifest.
//!
//! Both functions are READ-ONLY by construction (they take `&Session` and
//! never mutate; "snapshot is read-only" — AI-REQ-003). Field names are kept
//! short on purpose: this JSON is LLM context, so every byte costs tokens
//! (06_SCENE_SNAPSHOT_RESEARCH). Aliasing/truncation for the prompt happens in
//! the UI-side `ai/snapshot.ts` service; this layer returns stable truth.
//!
//! NOTHING secret ever flows through here: the snapshot contains document
//! structure only — no keys, no file paths, no clipboard, no pixels
//! (AI-REQ-002), and only the ACTIVE scene (per-document scoping).

use crate::model::{Document, Frame, Layer, LayerKind, LoopMode, Node, ShapeKind, SymbolType};
use crate::session::Session;
use serde::Serialize;
use std::collections::BTreeMap;

// ---------------------------------------------------------------------------
// E-AI-2 — scene snapshot
// ---------------------------------------------------------------------------

fn is_false(b: &bool) -> bool {
    !*b
}

#[derive(Serialize)]
struct SnapSettings {
    w: f64,
    h: f64,
    fps: u32,
    bg: String,
    #[serde(rename = "bgA")]
    bg_a: f64,
}

#[derive(Serialize)]
struct SnapScene {
    i: usize,
    name: String,
    count: usize,
}

#[derive(Serialize)]
struct SnapCounts {
    layers: usize,
    nodes: usize,
    keyframes: usize,
    tweens: usize,
    symbols: usize,
}

fn is_zero(v: &usize) -> bool {
    *v == 0
}

#[derive(Serialize)]
struct SnapKf {
    f: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    label: Option<String>,
    /// Blank frames CLEAR held content — the model must see them explicitly.
    #[serde(skip_serializing_if = "is_false")]
    blank: bool,
    /// Content id count (keyframes only); omitted when empty.
    #[serde(skip_serializing_if = "is_zero")]
    n: usize,
}

#[derive(Serialize)]
struct SnapTween {
    s: u32,
    e: u32,
    ease: f64,
}

#[derive(Serialize)]
struct SnapLayer {
    i: usize,
    id: u64,
    name: String,
    kind: &'static str,
    vis: bool,
    lock: bool,
    #[serde(skip_serializing_if = "is_false")]
    outline: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    oc: Option<String>,
    /// Parent folder's LAYER ID (organizational only).
    #[serde(skip_serializing_if = "Option::is_none")]
    parent: Option<u64>,
    #[serde(skip_serializing_if = "is_false")]
    collapsed: bool,
    kf: Vec<SnapKf>,
    tw: Vec<SnapTween>,
}

#[derive(Serialize, Default)]
struct SnapNode {
    id: u64,
    kind: String,
    /// Keyframes whose content includes this node: [layer_index, frame] pairs.
    kf: Vec<(usize, u32)>,
    // Shape rows ----------------------------------------------------------
    #[serde(skip_serializing_if = "Option::is_none")]
    x: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    y: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    sx: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    sy: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    r: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    w: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    h: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    fill: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    stroke: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    sw: Option<f64>,
    // Symbol rows ---------------------------------------------------------
    #[serde(skip_serializing_if = "Option::is_none")]
    sym: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    lp: Option<&'static str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    ff: Option<u32>,
}

#[derive(Serialize)]
struct SnapSymbol {
    id: u64,
    name: String,
    #[serde(rename = "type")]
    symbol_type: &'static str,
    uses: u32,
    dur: u32,
}

#[derive(Serialize)]
struct SnapRoot {
    /// Snapshot format marker (bump on shape change; UI parser refuses others).
    v: u32,
    /// E-AI-4 revision at build time (staleness detection).
    rev: u64,
    settings: SnapSettings,
    scene: SnapScene,
    active_layer: usize,
    playhead: u32,
    duration: u32,
    selection: Vec<u64>,
    counts: SnapCounts,
    layers: Vec<SnapLayer>,
    nodes: Vec<SnapNode>,
    library: Vec<SnapSymbol>,
}

fn shape_kind_name(k: ShapeKind) -> &'static str {
    // Exhaustive ON PURPOSE: adding a ShapeKind variant breaks this match —
    // the compiler forces this manifest/snapshot to stay truthful (single
    // source of truth; AI-REQ-111).
    match k {
        ShapeKind::Rect => "rect",
        ShapeKind::Oval => "oval",
    }
}

fn symbol_type_name(t: &SymbolType) -> &'static str {
    match t {
        SymbolType::Graphic => "graphic",
        SymbolType::MovieClip => "movieclip",
        SymbolType::Button => "button",
    }
}

fn loop_mode_name(m: &LoopMode) -> &'static str {
    match m {
        LoopMode::Loop => "loop",
        LoopMode::PlayOnce => "once",
        LoopMode::SingleFrame => "single",
    }
}

fn layer_row(i: usize, l: &Layer) -> SnapLayer {
    let kf = l
        .keyframes
        .iter()
        .map(|(f, fr)| SnapKf {
            f: *f,
            label: fr.label().map(str::to_string),
            blank: matches!(fr, Frame::Blank),
            n: match fr {
                Frame::Keyframe { content, .. } => content.len(),
                Frame::Blank => 0,
            },
        })
        .collect();
    let tw = l
        .tweens
        .iter()
        .map(|(s, t)| SnapTween {
            s: *s,
            e: t.end,
            ease: t.ease,
        })
        .collect();
    SnapLayer {
        i,
        id: l.id.0,
        name: l.name.clone(),
        kind: match l.kind {
            LayerKind::Normal => "normal",
            LayerKind::Folder => "folder",
        },
        vis: l.visible,
        lock: l.locked,
        outline: l.outline,
        oc: if l.outline { Some(l.outline_color.clone()) } else { None },
        parent: l.parent_id.map(|p| p.0),
        collapsed: l.collapsed,
        kf,
        tw,
    }
}

/// Compact semantic snapshot of the ACTIVE scene (Tier 0+1 of
/// 06_SCENE_SNAPSHOT_RESEARCH). `""` only when the active scene is missing.
pub fn scene_snapshot(session: &Session) -> String {
    let doc = &session.doc;
    let Some(scene) = doc.scene(session.active_scene) else {
        return "{}".to_string();
    };

    // Node membership: which keyframes (layer index, frame) reference each id.
    let mut membership: BTreeMap<u64, Vec<(usize, u32)>> = BTreeMap::new();
    let mut keyframe_count = 0usize;
    let mut tween_count = 0usize;
    for (li, layer) in scene.layers.iter().enumerate() {
        tween_count += layer.tweens.len();
        for (f, fr) in &layer.keyframes {
            keyframe_count += 1;
            if let Frame::Keyframe { content, .. } = fr {
                for id in content {
                    membership.entry(id.0).or_default().push((li, *f));
                }
            }
        }
    }

    let mut nodes: Vec<SnapNode> = Vec::with_capacity(membership.len());
    for (id, kf) in &membership {
        let Some(node) = doc.nodes.get(&crate::id::NodeId(*id)) else {
            continue;
        };
        let mut row = SnapNode {
            id: *id,
            kind: String::new(),
            kf: kf.clone(),
            ..SnapNode::default()
        };
        match node {
            Node::Rect {
                transform,
                width,
                height,
                fill,
                stroke,
                stroke_width,
                shape,
                ..
            } => {
                row.kind = shape_kind_name(*shape).to_string();
                row.x = Some(transform.x);
                row.y = Some(transform.y);
                row.sx = Some(transform.scale_x);
                row.sy = Some(transform.scale_y);
                row.r = Some(transform.rotation);
                row.w = Some(*width);
                row.h = Some(*height);
                row.fill = Some(fill.clone());
                row.stroke = stroke.clone();
                row.sw = Some(*stroke_width);
            }
            Node::SymbolInstance {
                transform,
                symbol_id,
                loop_mode,
                first_frame,
                ..
            } => {
                row.kind = "symbol".to_string();
                row.x = Some(transform.x);
                row.y = Some(transform.y);
                row.sx = Some(transform.scale_x);
                row.sy = Some(transform.scale_y);
                row.r = Some(transform.rotation);
                row.sym = Some(symbol_id.0);
                row.lp = Some(loop_mode_name(loop_mode));
                row.ff = Some(*first_frame);
            }
        }
        nodes.push(row);
    }

    let root = SnapRoot {
        v: 1,
        rev: session.doc_revision(),
        settings: SnapSettings {
            w: doc.settings.width,
            h: doc.settings.height,
            fps: doc.settings.fps,
            bg: doc.settings.background.clone(),
            bg_a: doc.settings.background_alpha,
        },
        scene: SnapScene {
            i: session.active_scene,
            name: scene.name.clone(),
            count: doc.scenes.len(),
        },
        active_layer: session.active_layer,
        playhead: session.playhead,
        duration: doc.timeline_duration(session.active_scene),
        selection: session.selection.iter().map(|id| id.0).collect(),
        counts: SnapCounts {
            layers: scene.layers.len(),
            nodes: nodes.len(),
            keyframes: keyframe_count,
            tweens: tween_count,
            symbols: doc.library.len(),
        },
        layers: scene
            .layers
            .iter()
            .enumerate()
            .map(|(i, l)| layer_row(i, l))
            .collect(),
        nodes,
        library: doc
            .library
            .iter()
            .map(|s| SnapSymbol {
                id: s.id.0,
                name: s.name.clone(),
                symbol_type: symbol_type_name(&s.symbol_type),
                uses: doc.symbol_use_count(s.id),
                dur: s.duration(),
            })
            .collect(),
    };

    serde_json::to_string(&root).unwrap_or_else(|_| "{}".to_string())
}

// ---------------------------------------------------------------------------
// E-AI-5 — trusted runtime capability manifest
// ---------------------------------------------------------------------------

/// Enumerate shape kinds THROUGH the exhaustive matcher so a new variant is a
/// compile error here until the manifest is updated (the single source of
/// truth trick — AI-REQ-111: no second hand-maintained AI capability list).
fn all_shape_kinds() -> Vec<&'static str> {
    [ShapeKind::Rect, ShapeKind::Oval]
        .iter()
        .map(|k| shape_kind_name(*k))
        .collect()
}

/// The engine's honest capability manifest. `features` flags flip ONLY when the
/// corresponding engine increment actually lands; anything absent-or-false is
/// treated as unavailable by the AI registry downstream (spec 07 manifest v0
/// mirrors 02_CURRENT_ENGINE_AUDIT exactly).
pub fn capabilities() -> String {
    let manifest = serde_json::json!({
        "v": 1,
        "engine": "kineora-core",
        "manifestFormat": "kineora-ai-manifest",
        // Shape kinds the shape pipeline can draw/hit-test/export today.
        "shapes": all_shape_kinds(),
        // Node families: "shape" = parametric rect/oval nodes; "symbol" =
        // placed symbol instances. (text/path/bitmap families do not exist.)
        "nodeFamilies": ["shape", "symbol"],
        "features": {
            "classicTween": true,      // Part 09.5 spans, numeric ease only
            "perKeyTransform": true,   // per-keyframe transform overrides
            "symbols": true,           // convert/create/place/swap/loop/delete
            "folders": true,           // F-20-05 layer folders w/ guards
            "instanceLoopModes": true, // graphic loop/play-once/single-frame
            "scenes": true,            // create scene (switch/delete pending)
            "frameLabels": true,       // Part 07 §7.2
            "arrangeAlign": true,      // arrange + align ops
            "strokeAtDraw": true,      // E1: stroke params on draw
            "selectionByIds": true,    // E-AI-3 (this build)
            "compositeUndo": true,     // E-AI-1 (this build)
            "nodeOpacity": false,      // no per-node alpha field (audit Q8)
            "namedEasings": false,     // Penner lib exists, NOT wired to tweens
            "paths": false,            // PATH model pending (tools lane E2)
            "text": false,             // no text node kind
            "motionTween": false,
            "shapeTween": false,
            "masks": false,
            "camera": false,
            "audio": false,
        },
    });
    serde_json::to_string(&manifest).unwrap_or_else(|_| "{}".to_string())
}

/// Kept for future per-document manifest extensions (e.g. platform-specific
/// capability trims) — today the manifest is engine-global.
pub fn capabilities_for(_doc: &Document) -> String {
    capabilities()
}
